### ISSUE TITLE
P2WSH address / script generation accepts PublicKey where compressed = false

### ISSUE THREAD

**Issue by junderw** [rust-bitcoin/rust-bitcoin#433]
Related: #427 #428 

All public keys in the witness stack are required to be compressed.

I can generate a P2WSH address and script by passing an uncompressed key.

Methods with the problem:

- `Script.to_v0_p2wsh()`
- `Address::p2wsh()`

Example:

```
[dependencies]
rand = "0.6"
bitcoin = { version = "0.23.0", features = ["rand"] }
```

```rust
use bitcoin::blockdata;
use bitcoin::network::constants::Network::Bitcoin;
use bitcoin::secp256k1::Secp256k1;
use bitcoin::util::key;
use bitcoin::Address;
use rand::rngs::OsRng;

struct Keypair {
    private: key::PrivateKey,
    public: key::PublicKey,
}

fn get_new_key() -> Keypair {
    let secp = Secp256k1::new();
    let mut rng = OsRng::new().expect("OsRng");
    let (secret_key, _) = secp.generate_keypair(&mut rng);
    let private = key::PrivateKey {
        compressed: false,
        key: secret_key,
        network: Bitcoin,
    };
    let public = private.public_key(&secp);

    return Keypair { private, public };
}

fn main() {
    let keypair = get_new_key();
    assert!(keypair.private.compressed == false);
    assert!(keypair.public.compressed == false);

    let pub_bytes = &keypair.public.to_bytes()[..];
    // This is uncompressed
    assert!(pub_bytes.len() == 65);

    // Make a P2PK manually
    let uncompressed_p2pk = blockdata::script::Builder::new()
        .push_slice(&pub_bytes)
        .push_opcode(blockdata::opcodes::all::OP_CHECKSIG)
        .into_script();

    // Should error but doesn't... SegWit can't take uncompressed pubkey
    let p2wsh1 = uncompressed_p2pk.to_v0_p2wsh();
    // Should error but doesn't... SegWit can't take uncompressed pubkey
    let address = Address::p2wsh(&uncompressed_p2pk, Bitcoin);

    println!(
        "Oops...\nWe got\n{:?}\nand\n{:?}\nwithout errors...",
        p2wsh1, address
    );
}
```

**Comment by apoelstra** [comment]
Related https://github.com/apoelstra/rust-miniscript/pull/88

I wonder if we want to use the type system here - change `PublicKey` to `LegacyPublicKey` and have a new `PublicKey` which is compressed-only. Then refuse to convert `LegacyPublicKey`s to segwit scripts.

**Comment by junderw** [comment]
I think your comment is for #427 (And a proposed fix #428 ) which is for P2WPKH.

This issue is for P2WSH.

**Comment by apoelstra** [comment]
I think at a high level these are all the same issue - Segwit does not allow uncompressed keys but rust-bitcoin (and associated crates) makes no effort to enforce this.

**Comment by junderw** [comment]
True.

I am new to rust, but is it unnatural for someone building a script for p2wsh conversion to only use [u8] and not go through `PublicKey` at all?

Usually in JS a lot of people use hex strings etc. convert to Buffer.

If it would be nearly impossible or highly unnatural for any use case to NOT pass through LegacyPublicKey before entering the p2wsh methods, then that solves it imo.

It seems like someone trying to import a script from some API or database where it's stored in hex would convert to [u8] then create a Script without the Builder.

Correct me if I'm wrong.

**Comment by stevenroose** [comment]
I think the issue here is that all script-hash based methods like `Address::p2wsh` take a script "type" and not a public key type. So you can't enforce this using the type system. Only by parsing into the `Script` and that's quite messy.

**Comment by junderw** [comment]
Perhaps a Script type should be basically a vector of `op_code OR pubkey OR signature OR [u8]` So basically you would encourage people to use the types for pubkeys and sigs, and the only way to get around the compressed check would be to fiddle with encoding to get the pubkey as a [u8].

Just a thought... not quite sure the implications throughout the API tho.

**Comment by junderw** [comment]
Super rough idea. Instead of being a wrapper around [u8] it could be a wrapper around Vec<ScriptItem> (below) etc.

```rust
enum ScriptItem {
    PublicKey(PublicKey),
    Signature(Signature),
    OpCode(u8),
    Bytes([u8]),
}

fn main() {
    let mut output: Vec<ScriptItem> = Vec::new();
    output.push(ScriptItem::Signature(sig));
    output.push(ScriptItem::PublicKey(pubkey));
    output.push(ScriptItem::OpCode(op_code::OP_CHECKSIG));
}
```

**Comment by junderw** [comment]
Any thoughts on my idea [above](https://github.com/rust-bitcoin/rust-bitcoin/issues/433#issuecomment-650920893)

**Comment by Kixunil** [comment]
I think you could still screw it up by returning bytes. If I understand the issue correctly this whole thing is just general danger of working with (raw) scripts?

**Comment by junderw** [comment]
> I think you could still screw it up by returning bytes.

Yes. But by documenting that a script should be built with the enum, and you should NOT use pubkey or signature byte slices directly in `ScriptItem::Bytes` you can prevent a large portion of foot guns.

Not to mention that currently the segwit compressed pubkey constraint is mentioned nowhere in the source or documentation. By adding this enum and a check that any Pubkeys are compressed, we have it written in code. We can also add it in the documentation for script as well (to explain the enum usage)

---

In bitcoinjs, we have a validation step that just throws an error if there's a 65 byte push that starts with 0x04. And we have a method for disabling validation altogether, so someone who reeeeeally needs a 65 byte data point, or needs to hash a uncompressed pubkey without using any CHECKSIG ops or something can disable validation for that one script.

We realized this because a user created a p2wsh multisig using uncompressed pubkeys.

**Comment by Kixunil** [comment]
Hmm, that sounds pretty convincing. Although I wonder if just having a more straightforward API for these kinds of things would be better. What do others think?

**Comment by apoelstra** [comment]
I'm generally pretty hesitant to make `Script` richer, because solutions like this don't go all the way (and if they did, you'd wind up with rust-miniscript, which is significantly larger than this entire library).

We do have the `push_key` method on `Script` right now. I wonder if this could be modified with a boolean flag, or with a type parameter, to disallow uncompressed keys in segwit contexts? We do this sort of thing throughout rust-miniscript. Sometimes it works cleanly and sometimes it's a huge mess which winds up being a 10000 line diff to implement and then we have to revert it. It's hard to tell in advance.
